AsyncDisplayKit/Layout/ASDimension.h

```diff
@@ -12,56 +12,183 @@
 #import <UIKit/UIKit.h>
 #import <AsyncDisplayKit/ASBaseDefines.h>
 
-/** A dimension relative to constraints to be provided in the future. */
-typedef NS_ENUM(NSInteger, ASRelativeDimensionType) {
+#define ASPointsAreValidForLayout(x) ((isnormal(x) || x == 0.0) && x >= 0.0 && x < (CGFLOAT_MAX / 2.0))
+#define ASCGSizeIsValidForLayout(x) (ASPointsAreValidForLayout(x.width) && ASPointsAreValidForLayout(x.height))
+
+/**
+ * A dimension relative to constraints to be provided in the future.
+ * A RelativeDimension can be one of three types:
+ *
+ * "Auto" - This indicated "I have no opinion" and may be resolved in whatever way makes most sense given the circumstances.
+ *
+ * "Points" - Just a number. It will always resolve to exactly this amount.
+ *
+ * "Percent" - Multiplied to a provided parent amount to resolve a final amount.
+ */
+typedef NS_ENUM(NSInteger, ASDimensionType) {
+  /** This indicates "I have no opinion" and may be resolved in whatever way makes most sense given the circumstances. */
+  ASDimensionTypeAuto,
   /** Just a number. It will always resolve to exactly this amount. This is the default type. */
-  ASRelativeDimensionTypePoints,
+  ASDimensionTypePoints,
   /** Multiplied to a provided parent amount to resolve a final amount. */
-  ASRelativeDimensionTypeFraction,
+  ASDimensionTypeFraction,
 };
 
 typedef struct {
-  ASRelativeDimensionType type;
+  ASDimensionType type;
   CGFloat value;
-} ASRelativeDimension;
+} ASDimension;
+
+/**
+ * Expresses a size with relative dimensions.
+ */
+typedef struct {
+  ASDimension width;
+  ASDimension height;
+} ASRelativeSize;
```

Comment:
I moved the declaration of `ASRelativeSize` into `ASDimension.m` as we use it only internally for a bunch of calculations